File: 0120-triangle/0120-triangle.py
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        def findmin(i, j):
            if i == len(triangle)-1:
                return triangle[i][j]
            if memo[i][j]==-1:
                memo[i][j] = triangle[i][j] + min(findmin(i+1, j), findmin(i+1, j+1))
            return memo[i][j]
        
        memo = [[-1 for j in range(i+1)] for i in range(len(triangle))]
        return findmin(0, 0)
        

File: 0120-triangle/0120-triangle.java
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        int l=triangle.size();
        int[][] dp = new int[l][triangle.get(l-1).size()];
        for(int[] row: dp)
            Arrays.fill(row, -1);
        return helper(triangle, 0, 0, dp);
    }
    int helper(List<List<Integer>> list, int row, int col, int[][] dp){
        if(row==list.size()-1){
            dp[row][col]=list.get(row).get(col);
            return dp[row][col];
        }
        if(dp[row][col] == -1)
            dp[row][col] = Math.min(helper(list, row+1, col, dp), helper(list, row+1, col+1, dp))
                + list.get(row).get(col);
        return dp[row][col];
    }
}